engine/src/main/java/org/pentaho/di/repository/kdr/delegates/KettleDatabaseRepositoryUserDelegate.java

```diff
@@ -0,0 +1,181 @@
+import org.pentaho.di.i18n.BaseMessages;
+import org.pentaho.di.repository.IUser;
+import org.pentaho.di.repository.ObjectId;
+import org.pentaho.di.repository.RepositoryElementInterface;
+import org.pentaho.di.repository.UserInfo;
+import org.pentaho.di.repository.kdr.KettleDatabaseRepository;
+
+public class KettleDatabaseRepositoryUserDelegate extends KettleDatabaseRepositoryBaseDelegate {
+
+  private static Class<?> PKG = UserInfo.class; // for i18n purposes, needed by Translator2!!
+
+  public KettleDatabaseRepositoryUserDelegate( KettleDatabaseRepository repository ) {
+    super( repository );
+  }
+
+  public RowMetaAndData getUser( ObjectId id_user ) throws KettleException {
+    return repository.connectionDelegate.getOneRow(
+      quoteTable( KettleDatabaseRepository.TABLE_R_USER ), quote( KettleDatabaseRepository.FIELD_USER_ID_USER ),
+      id_user );
+  }
+
+  public synchronized ObjectId getUserID( String login ) throws KettleException {
+    return repository.connectionDelegate.getIDWithValue(
+      quoteTable( KettleDatabaseRepository.TABLE_R_USER ), quote( KettleDatabaseRepository.FIELD_USER_ID_USER ),
+      quote( KettleDatabaseRepository.FIELD_USER_LOGIN ), login );
+  }
+
+  // Load user with login from repository, don't verify password...
+  public IUser loadUserInfo( IUser userInfo, String login ) throws KettleException {
+    try {
+      userInfo.setObjectId( getUserID( login ) );
+      if ( userInfo.getObjectId() != null ) {
+        RowMetaAndData r = getUser( userInfo.getObjectId() );
+        if ( r != null ) {
+          userInfo.setLogin( r.getString( "LOGIN", null ) );
+          userInfo.setPassword( Encr.decryptPassword( r.getString( "PASSWORD", null ) ) );
+          userInfo.setUsername( r.getString( "NAME", null ) );
+          userInfo.setDescription( r.getString( "DESCRIPTION", null ) );
+          userInfo.setEnabled( r.getBoolean( "ENABLED", false ) );
+          return userInfo;
+        } else {
+          throw new KettleDatabaseException( BaseMessages.getString( PKG, "UserInfo.Error.UserNotFound", login ) );
+        }
+      } else {
+        throw new KettleDatabaseException( BaseMessages.getString( PKG, "UserInfo.Error.UserNotFound", login ) );
+      }
+    } catch ( KettleDatabaseException dbe ) {
+      throw new KettleException( BaseMessages.getString( PKG, "UserInfo.Error.UserNotLoaded", login, "" ), dbe );
+    }
+  }
+
+  /**
+   * Load user with login from repository and verify the password...
+   *
+   * @param rep
+   * @param login
+   * @param passwd
+   * @throws KettleException
+   */
+  public IUser loadUserInfo( IUser userInfo, String login, String passwd ) throws KettleException {
+    if ( userInfo == null || login == null || login.length() <= 0 ) {
+      throw new KettleDatabaseException( BaseMessages.getString( PKG, "UserInfo.Error.IncorrectPasswortLogin" ) );
+    }
+
+    try {
+      loadUserInfo( userInfo, login );
+    } catch ( KettleException ke ) {
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

**Problem**: While wrapping the caught exception into a custom one, information about the caught exception is being lost, including information about the stack trace of the exception.

**Fix**: If the caught exception object does not contain sensitive information, consider passing it as the &#34;rootCause&#34; or inner exception parameter to the constructor of the new exception before throwing the new exception. (Note that not all exception constructors support inner exceptions. Use a wrapper exception that supports inner exceptions.) 
[Learn more](https://www.ibm.com/support/pages/best-practice-catching-and-re-throwing-java-exceptions)